File: protocol/serial-control.py
import sys
import serial
import time

# Constants based on the protocol
UART_CONTROL = 0xF8
UART_ENABLE = 0x08
COMMAND_START = 0xF7
MODE_BYTE = 0x34

def init_serial(portname):
    try:
        # Open the serial port with the specified settings
        ser = serial.Serial(
            port=portname,
            baudrate=9600,
            bytesize=serial.EIGHTBITS,
            parity=serial.PARITY_NONE,
            stopbits=serial.STOPBITS_ONE,
            timeout=1  # Timeout in seconds for read
        )
        if ser.is_open:
            print(f"Serial port {portname} opened successfully.")
        return ser
    except serial.SerialException as e:
        print(f"Error opening serial port {portname}: {e}")
        return None

def send_command(ser, energy, status_flags):
    # Construct the command byte array
    command = bytearray([
        UART_CONTROL,
        UART_ENABLE,
        COMMAND_START,
        MODE_BYTE,
        energy,         # Energy byte
        status_flags    # Status flags byte
    ])
    
    # Write the command to the serial port
    ser.write(command)
    print("Sent command sequence to device.")

def main():
    if len(sys.argv) != 4:
        print(f"Usage: {sys.argv[0]} <serial_port> <energy> <status_flags>")
        sys.exit(1)

    # Parse the arguments
    portname = sys.argv[1]
    energy = int(sys.argv[2], 0)        # Convert energy argument to an integer
    status_flags = int(sys.argv[3], 0)  # Convert status_flags argument to an integer

    # Initialize the serial port
    ser = init_serial(portname)
    if not ser:
        sys.exit(1)

    try:
        # Send the command to the device
        send_command(ser, energy, status_flags)
    finally:
        # Ensure the serial port is closed after use
        ser.close()
        print("Serial port closed.")

if __name__ == "__main__":
    main()

File: control/protocol/serial-control.c
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <string.h>
#include <unistd.h>
#include <fcntl.h>
#include <termios.h>

#define UART_CONTROL 0xF8
#define UART_ENABLE 0x08
#define COMMAND_START 0xF7
#define MODE_BYTE 0x34

// Function to initialize the serial port
int init_serial(const char *portname) {
    int fd = open(portname, O_RDWR | O_NOCTTY | O_SYNC);
    if (fd < 0) {
        perror("Error opening serial port");
        return -1;
    }

    struct termios tty;
    memset(&tty, 0, sizeof(tty));
    if (tcgetattr(fd, &tty) != 0) {
        perror("Error from tcgetattr");
        close(fd);
        return -1;
    }

    // Set baud rate
    cfsetospeed(&tty, B9600);
    cfsetispeed(&tty, B9600);

    // Set up 8N1 (8 data bits, no parity, 1 stop bit)
    tty.c_cflag = (tty.c_cflag & ~CSIZE) | CS8;
    tty.c_iflag &= ~IGNBRK; // disable break processing
    tty.c_lflag = 0;        // no signaling chars, no echo, no canonical processing
    tty.c_oflag = 0;        // no remapping, no delays
    tty.c_cc[VMIN]  = 1;    // read blocks until at least 1 char arrives
    tty.c_cc[VTIME] = 5;    // 0.5 seconds read timeout

    tty.c_iflag &= ~(IXON | IXOFF | IXANY); // shut off xon/xoff ctrl
    tty.c_cflag |= (CLOCAL | CREAD);        // ignore modem controls, enable reading
    tty.c_cflag &= ~(PARENB | PARODD);      // shut off parity
    tty.c_cflag &= ~CSTOPB;
    tty.c_cflag &= ~CRTSCTS;

    if (tcsetattr(fd, TCSANOW, &tty) != 0) {
        perror("Error from tcsetattr");
        close(fd);
        return -1;
    }
    return fd;
}

// Function to send command sequence
void send_command(int fd, uint8_t energy, uint8_t status_flags) {
    uint8_t buffer[6];
    
    // Initialize control and enable UART
    buffer[0] = UART_CONTROL;
    buffer[1] = UART_ENABLE;

    // Command sequence with energy and status flags
    buffer[2] = COMMAND_START;
    buffer[3] = MODE_BYTE;
    buffer[4] = energy;       // Energy byte at index 0x11
    buffer[5] = status_flags; // Status flags byte at index 0x51

    // Send the buffer to the serial port
    write(fd, buffer, sizeof(buffer));
    printf("Sent command sequence to device.\n");
}

int main(int argc, char *argv[]) {
    if (argc != 3) {
        fprintf(stderr, "Usage: %s <energy> <status_flags>\n", argv[0]);
        return EXIT_FAILURE;
    }

    // Parse energy and status flags from arguments
    uint8_t energy = (uint8_t)strtoul(argv[1], NULL, 0);
    uint8_t status_flags = (uint8_t)strtoul(argv[2], NULL, 0);

    // Initialize the serial port
    const char *portname = "/dev/ttyS0"; // Replace with your serial port
    int fd = init_serial(portname);
    if (fd < 0) {
        return EXIT_FAILURE;
    }

    // Send command to the device
    send_command(fd, energy, status_flags);

    // Close the serial port
    close(fd);
    return EXIT_SUCCESS;
}

